docs/adr/0002-yarn-v4.md

```diff
@@ -0,0 +1,175 @@
+### pnpDataPath is no longer configurable
+
+This used to be a configurable key in the `.yarnrc.yml` file. The previous default path, `./.pnp.data.json`, is now hard-coded and can't be changed.
+
+The only mention to `pnpDataPath` in Cachi2 is the [check](https://github.com/containerbuildsystem/cachi2/blob/a5f19c6f9be90be4289beee35ecccd2827bbb328/cachi2/core/package_managers/yarn/main.py#L43) for paths pointing outside of the repo. The check can be kept the same way so Yarn v3 can still be covered, and it will simply be skipped in v4.
+
+### Yarn now caches npm version metadata
+
+Although Yarn only seems to create the cache folder (`{globalFolder}/metadata/npm`) when the [hardened mode](https://yarnpkg.com/configuration/yarnrc#enableHardenedMode) is enabled. Simply running `yarn install` on a local project or even prefetching with Cachi2 doesn't seem to generate this extra folder.
+
+The metadata, in itself, is a collection of json files with a few kilobytes each. In case we decide to enable the hardened mode, it'd be ideal to delete this npm metadata this from the output folder, although there's no harm in keeping it.
+
+### Changes in yarn.lock
+
+When updating v3 projects to v4, some differences in the lockfile appear. These are some picks from updating the [disallowed-protocols]() branch on the cachi2-yarn-berry test project:
+
+**`npm:` locator added to npm dependencies:**
+```
+   dependencies:
+-    chownr: ^2.0.0
+-    fs-minipass: ^2.0.0
+-    minipass: ^5.0.0
+-    minizlib: ^2.1.1
+-    mkdirp: ^1.0.3
+-    yallist: ^4.0.0
+-  checksum: ...
++    chownr: "npm:^2.0.0"
++    fs-minipass: "npm:^2.0.0"
++    minipass: "npm:^5.0.0"
++    minizlib: "npm:^2.1.1"
++    mkdirp: "npm:^1.0.3"
++    yallist: "npm:^4.0.0"
++  checksum: ...
+```
+
+**Changes to some instances of the file locator:**
+```
+-  resolution: "strip-ansi-tarball@file:external-packages/strip-ansi-4.0.0.tgz::locator=berryscary%40workspace%3A."
++  resolution: "strip-ansi-tarball@file:external-packages/strip-ansi-4.0.0.tgz#external-packages/strip-ansi-4.0.0.tgz::hash=e17689&locator=berryscary%40workspace%3A."
+```
+
+Yarn v4 introduced the subdirectory (`#external-packages/strip-ansi-4.0.0.tgz`) and the hash (`::hash=e17689`), which are already [handled](https://github.com/containerbuildsystem/cachi2/blob/a5f19c6f9be90be4289beee35ecccd2827bbb328/cachi2/core/package_managers/yarn/locators.py#L77-L79) by the v3 implementation in Cachi2.
+
+**Changes to some instances of the patch locator:**
+```
+-  resolution: "typescript@patch:typescript@npm%3A5.1.6#~builtin<compat/typescript>::version=5.1.6&hash=5da071"
++  resolution: "typescript@patch:typescript@npm%3A5.1.6#optional!builtin<compat/typescript>::version=5.1.6&hash=5da071"
+```
+
+Instances of the patch locator are currently [being ignored](https://github.com/containerbuildsystem/cachi2/blob/a5f19c6f9be90be4289beee35ecccd2827bbb328/cachi2/core/package_managers/yarn/resolver.py#L264-L268) in the v3 implementation in Cachi2.
```

Comment:
Ooh, that seems like a bug. They're not being ignored, they're being reported as if they were regular registry dependencies. Which is not quite right, the patched dependency doesn't have to be a registry dependency.

Fixing the parsing of patch locators would be out of scope then? It would have to be fixed before cachi2 can start reporting the patches

Reply:
IIRC, we wanted to report patched dependencies as a regular dependency type and add the [pedigree.patches ](https://cyclonedx.org/docs/1.6/json/#components_items_pedigree_patches) information to the SBOM component.

I think the latter part is not being done, so you're right, we need to fix it.

Reply:
> IIRC, we wanted to report patched dependencies as a regular dependency type and add the [pedigree.patches ](https://cyclonedx.org/docs/1.6/json/#components_items_pedigree_patches) information to the SBOM component.

That makes sense, but the current approach has a bug: https://github.com/containerbuildsystem/cachi2/blob/d89b5674c55d48d33eeb9cf3acc6d77f8692f980/tests/integration/test_data/yarn_e2e_test/bom.json#L323

That's a purl for a regular registry dependency, generated for https://github.com/cachito-testing/cachi2-yarn-berry/blob/70515793108df42547d3320c7ea4cd6b6e505c46/yarn.lock#L272 (definitely not a registry dependency)

The correct purl also happens to be present, because yarn.lock also lists the "un-patched" dependency separately: https://github.com/containerbuildsystem/cachi2/blob/d89b5674c55d48d33eeb9cf3acc6d77f8692f980/tests/integration/test_data/yarn_e2e_test/bom.json#L335

So in practice, the bug is that cachi2 reports bogus registry dependencies in these situations